### PR TITLE
#17949: Make loopback address search more accurate

### PR DESCRIPTION
**Pull request by rl1987**
https://trac.torproject.org/projects/tor/ticket/17949

### CODE REVIEW

File: src/common/address.c [hunk 1]
```diff
@@ -1592,31 +1592,38 @@ ifreq_to_smartlist(char *buf, size_t buflen, int loopback)
...
-          continue;
-        smartlist_add(result, tor_memdup(&tmp, sizeof(tmp)));
-      }
+    int conversion_success = (tor_addr_from_sockaddr(&tmp, sa, NULL) == 0);
+    if (valid_sa_family && conversion_success) {
+      if (tor_addr_is_null(&tmp))
+        continue;
+      if (tor_addr_is_multicast(&tmp))
+        continue;
+#ifdef __linux__
```
Comment (isislovecruft) [inline src/common/address.c]: Does this part fail on OSX or Windows?

Comment (rl1987) [reply src/common/address.c]: No, but it's not needed on macOS as I can rely on flag inside `struct ifreq` to check if it's loopback interface. For some reason it's not working out on Linux. On Windows this entire function is not compiled in as Windows does not have `ioctl()`.

File: src/common/address.c [hunk 2]
```diff
@@ -1762,75 +1820,27 @@ get_interface_address6_via_udp_socket_hack,(int severity,
   return r;
 }
 
-/** Set *<b>addr</b> to an arbitrary IP address (if any) of an interface that
```
Comment (teor2345) [inline src/common/address.c]: Please don't move functions, it makes them harder to review.

If you want to move functions, do it in a separate commit, and make that commit the first or last commit of the branch. (On GitHub, please make a separate pull request.)

File: src/common/address.h [hunk 3]
```diff
@@ -206,13 +206,22 @@ const char * fmt_addr32(uint32_t addr);
 
 MOCK_DECL(int,get_interface_address6,(int severity, sa_family_t family,
 tor_addr_t *addr));
+MOCK_DECL(int,get_loopback_address6,(int severity, sa_family_t family,
+tor_addr_t *addr));
 void interface_address6_list_free_(smartlist_t * addrs);// XXXX
+
+void free_interface_address6_list(smartlist_t * addrs);
```
Comment (teor2345) [inline src/common/address.h]: Where is this function defined?
Why can't you use interface_address6_list_free()?

File: src/common/address.h [hunk 4]
```diff
@@ -327,9 +336,18 @@ char *tor_dup_ip(uint32_t addr) ATTR_MALLOC;
...
+{
+  free_interface_address6_list(addrs);
+}
+
 /** Return a smartlist of the IPv4 addresses of all interfaces on the server.
- * Excludes loopback and multicast addresses. Only includes internal addresses
- * if include_internal is true. (Note that a relay behind NAT may use an
+ * Excludes loopback and multicast addresses.
+ * Only includes internal addresses if <b>include_internal</b> is true.
+ * (Note that a relay behind NAT may use an
```
Comment (teor2345) [inline src/common/address.h]: In future, please make white space changes in a separate commit.

File: src/common/address.h [hunk 5]
```diff
@@ -327,9 +336,18 @@ char *tor_dup_ip(uint32_t addr) ATTR_MALLOC;
 MOCK_DECL(int,get_interface_address,(int severity, uint32_t *addr));
 #define interface_address_list_free(lst)\
   interface_address6_list_free(lst)
+/** Free a smartlist of IP addresses returned by get_interface_address_list.
+ */
+static inline void
```
Comment (teor2345) [inline src/common/address.h]: Please follow the convention: interface_address_list_free, not free_interface_address_list.

File: src/or/config.c [hunk 6]
```diff
@@ -2805,7 +2805,8 @@ resolve_my_address(int warn_severity, const or_options_t *options,
                "resolves to a private IP address (%s). Trying something "
                "else.", hostname, fmt_addr32(addr));
 
-        if (get_interface_address6(warn_severity, AF_INET, &interface_ip)<0) {
+        if (get_interface_address6(warn_severity, AF_INET,
+            &interface_ip) < 0) {
```
Comment (teor2345) [inline src/or/config.c]: In future, please make white space changes in a separate commit.

File: src/test/test_addr.c [hunk 7]
```diff
@@ -763,6 +763,8 @@ test_addr_ip6_helpers(void *arg)
   tt_int_op(r, OP_LE, 0); // "it worked or it didn't"
   i = get_interface_address6(LOG_DEBUG, AF_INET6, &t2);
   tt_int_op(i, OP_LE, 0); // "it worked or it didn't"
+  r = get_interface_address6(LOG_DEBUG, AF_INET, &t1);
+  i = get_interface_address6(LOG_DEBUG, AF_INET6, &t2);
```
Comment (teor2345) [inline src/test/test_addr.c]: Why add these extra unit tests?
They don't seem to do anything different to the lines above them.

File: src/test/test_address.c [hunk 8]
```diff
@@ -221,7 +221,7 @@ test_address_ifaddrs_to_smartlist(void *arg)
    ifa_ipv6->ifa_dstaddr = NULL;
    ifa_ipv6->ifa_data = NULL;
 
-   smartlist = ifaddrs_to_smartlist(ifa, AF_UNSPEC);
+   smartlist = ifaddrs_to_smartlist(ifa, AF_UNSPEC, 0);
```
Comment (teor2345) [inline src/test/test_address.c]: For every test that you added a "0" (non-loopback), please write a version of the test that tests "1" (loopback).

File: src/common/address.c [hunk 9]
```diff
@@ -1762,75 +1820,27 @@ get_interface_address6_via_udp_socket_hack,(int severity,
...
-MOCK_IMPL(smartlist_t *,
-get_interface_address6_list,(int severity,
-                             sa_family_t family,
-                             int include_internal))
+static smartlist_t *
+get_interface_address6_list_impl(int severity, sa_family_t family,
+                                 int include_internal, int loopback)
 {
   smartlist_t *addrs;
   tor_addr_t addr;
```
[the remark this answers is not in the document]
Comment (rl1987) [reply src/common/address.c]: Did so in 546ae9fbd066a2e17c2cf13a9b13ad775360beb0.

File: src/common/address.h (hunk 3 above)
Comment (rl1987) [reply src/common/address.h]: Turns out I use `interface_address6_list_free()` everywhere. Removed this in b8645a82ab0bbc7a8adb879024c1ea2e31f5f270.

File: src/common/address.h (hunk 5 above)
Comment (rl1987) [reply src/common/address.h]: Removed in b8645a82ab0bbc7a8adb879024c1ea2e31f5f270.

File: src/common/address.c [hunk 10]
```diff
@@ -1460,6 +1474,12 @@ ip_adapter_addresses_to_smartlist(const IP_ADAPTER_ADDRESSES *addresses)
         continue;
       if (tor_addr_from_sockaddr(&tmp, sa, NULL) < 0)
         continue;
+      if (loopback && !tor_addr_is_loopback(&tmp))
+        continue;
+      if (tor_addr_is_null(&tmp))
+        continue;
+      if (tor_addr_is_multicast(&tmp))
+        continue;
```
[the remark this answers is not in the document]
Comment (rl1987) [reply src/common/address.c]: Fixed in 782933e5e24d0f458ddf20724d640be3669eb644.

File: src/common/address.c [hunk 11]
```diff
@@ -1404,8 +1407,16 @@ ifaddrs_to_smartlist(const struct ifaddrs *ifa, sa_family_t family)
...
+    if (loopback && (i->ifa_flags & IFF_LOOPBACK) == 0)
+      continue;
+    if (!loopback && (i->ifa_flags & IFF_LOOPBACK) == IFF_LOOPBACK)
+      continue;
     if (tor_addr_from_sockaddr(&tmp, i->ifa_addr, NULL) < 0)
       continue;
+    if (tor_addr_is_null(&tmp))
+      continue;
+    if (tor_addr_is_multicast(&tmp))
+      continue;
```
[the remark this answers is not in the document]
Comment (rl1987) [reply src/common/address.c]: Fixed in 782933e5e24d0f458ddf20724d640be3669eb644.

File: src/common/address.c [hunk 12]
```diff
@@ -1472,7 +1492,8 @@ ip_adapter_addresses_to_smartlist(const IP_ADAPTER_ADDRESSES *addresses)
  * <b>tor_addr_t</b>  structures.
```
[the remark this answers is not in the document]
Comment (rl1987) [reply src/common/address.c]: Fixed in ae9dc56ae042d1a170e836e6eadbf5b69135e26a and 7aded770e5c00512f8b277b6754004f53e50f572.

File: src/common/address.c [hunk 13]
```diff
@@ -1690,12 +1731,14 @@ tor_addr_is_multicast(const tor_addr_t *a)
 /** Attempt to retrieve IP address of current host by utilizing some
  * UDP socket trickery. Only look for address of given <b>family</b>
  * (only AF_INET and AF_INET6 are supported). Set result to *<b>addr</b>.
- * Return 0 on success, -1 on failure.
+ * Return 0 on success, -1 on failure. Try to retrieve loopback interface
+ * address if <b>loopback</b> is true.
```
[the remark this answers is not in the document]
Comment (rl1987) [reply src/common/address.c]: Fixed - ae9dc56ae042d1a170e836e6eadbf5b69135e26a.

File: src/common/address.c [hunk 14]
```diff
@@ -1878,6 +1888,107 @@ get_interface_address6_list,(int severity,
...
+ */
+MOCK_IMPL(smartlist_t *,get_interface_address6_list,(int severity,
+                                                     sa_family_t family,
+                                                     int include_internal))
+{
+  return get_interface_address6_list_impl(severity, family, include_internal,
+                                          0);
+}
+
+MOCK_IMPL(smartlist_t *,get_loopback_address6_list,(int severity,
```
[the remark this answers is not in the document]
Comment (rl1987) [reply src/common/address.c]: Wrote a comment in 12e7e436fd77e2f48838d6613b1cd8d53b72bcf1.

File: src/common/address.c [hunk 15]
```diff
@@ -1878,6 +1888,107 @@ get_interface_address6_list,(int severity,
   return addrs;
 }
 
+/**
+ * If <b>loopback</b> is false, fetch a public network interface
```
[the remark this answers is not in the document]
Comment (rl1987) [reply src/common/address.c]: Fixed in a3668ddf550cff7f99003918ce5ab9657332ba77.

File: src/common/address.c [hunk 16]
```diff
@@ -1762,75 +1820,27 @@ get_interface_address6_via_udp_socket_hack,(int severity,
...
- * addresses.  This address should only be used in checking whether our
- * address has changed, as it may be an internal IP address.  Return 0 on
- * success, -1 on failure.
- * Prefer get_interface_address6_list for a list of all addresses on all
- * interfaces which connect to the Internet.
+/**
+ * Return a list of <b>tor_addr_t</b> objects with network interface
+ * addresses that we find on the system. If <b>include_internal</b>
+ * is true, include non public addresses (e.g. for Tor instances
+ * running behind NAT).
```
[the remark this answers is not in the document]
Comment (rl1987) [reply src/common/address.c]: Fixed in 82681a1968fb3b3c4964811ff6bd21291ec5b06f.

File: src/common/address.c [hunk 17]
```diff
@@ -1747,9 +1795,19 @@ get_interface_address6_via_udp_socket_hack,(int severity,
     goto err;
   }
 
-  if (tor_addr_is_loopback(addr) || tor_addr_is_multicast(addr)) {
+  if ((!loopback && tor_addr_is_loopback(addr))
+      || tor_addr_is_multicast(addr)) {
     log_fn(severity, LD_NET, "Address that we determined via UDP socket"
-           " magic is unsuitable for public comms.");
+                             " fallback technique is unusable by Tor.");
```
[the remark this answers is not in the document]
Comment (rl1987) [reply src/common/address.c]: In 3d56e7fe0482c213f4634109c37c8dd7eb9ff3bd I deleted this entire check as I'm checking for null and multicast properties in `_list` now (see 782933e5e24d0f458ddf20724d640be3669eb644).

File: src/common/address.c [hunk 18]
```diff
@@ -1553,10 +1574,11 @@ ifconf_free_ifc_buf(struct ifconf *ifc)
 }
 
 /** Convert <b>*buf</b>, an ifreq structure array of size <b>buflen</b>,
- * into smartlist of <b>tor_addr_t</b> structures.
+ * into smartlist of <b>tor_addr_t</b> structures. Skip non-loopback
+ * addresses if <b>loopback</b> is true.
```
[the remark this answers is not in the document]
Comment (rl1987) [reply src/common/address.c]: Fixed.

File: src/common/address.c [hunk 19]
```diff
@@ -1569,18 +1591,33 @@ ifreq_to_smartlist(char *buf, size_t buflen)
...
+      int valid_sa_family = (sa->sa_family == AF_INET ||
+                             sa->sa_family == AF_INET6);
+
+      int conversion_success = (tor_addr_from_sockaddr(&tmp, sa, NULL) == 0);
+
+      if (valid_sa_family && conversion_success) {
+        if (tor_addr_is_null(&tmp))
+          continue;
+        if (tor_addr_is_multicast(&tmp))
+          continue;
```
[the remark this answers is not in the document]
Comment (rl1987) [reply src/common/address.c]: Did so in 782933e5e24d0f458ddf20724d640be3669eb644 and 3d56e7fe0482c213f4634109c37c8dd7eb9ff3bd.

File: src/common/address.c [hunk 20]
```diff
@@ -1589,10 +1626,12 @@ ifreq_to_smartlist(char *buf, size_t buflen)
 
 /** Use ioctl(.,SIOCGIFCONF,.) to get a list of current machine
  * network interface addresses. Represent the result by smartlist of
- * <b>tor_addr_t</b> structures.
+ * <b>tor_addr_t</b> structures. Skip non-loopback addresses if
+ * <b>loopback</b> is true.
```
[the remark this answers is not in the document]
Comment (rl1987) [reply src/common/address.c]: Fixed.

File: src/common/address.c [hunk 21]
```diff
@@ -1747,9 +1795,19 @@ get_interface_address6_via_udp_socket_hack,(int severity,
...
   }
 
-  if (tor_addr_is_loopback(addr) || tor_addr_is_multicast(addr)) {
+  if ((!loopback && tor_addr_is_loopback(addr))
+      || tor_addr_is_multicast(addr)) {
     log_fn(severity, LD_NET, "Address that we determined via UDP socket"
-           " magic is unsuitable for public comms.");
+                             " fallback technique is unusable by Tor.");
+  } else if (!loopback && tor_addr_is_loopback(addr)) {
+    log_fn(severity, LD_NET, "Tried to get public adddress via UDP socket"
```
[the remark this answers is not in the document]
Comment (rl1987) [reply src/common/address.c]: Deleted in 431d310285ee413a075726f0cdf353f0d51cf832.

File: src/common/address.c [hunk 22]
```diff
@@ -1747,9 +1795,19 @@ get_interface_address6_via_udp_socket_hack,(int severity,
...
+                             " fallback technique, but got a loopback address.");
+  } else if (loopback && !tor_addr_is_loopback(addr)) {
+    log_fn(severity, LD_NET, "Tried to perform an UDP fallback technique to "
+                             "get loopback interface address, but "
+                             "got non-loopback address instead. Your "
+                             "loopback configuration might not be "
+                             "standard.");
```
[the remark this answers is not in the document]
Comment (rl1987) [reply src/common/address.c]: Deleted this entire check in 431d310285ee413a075726f0cdf353f0d51cf832. However, I am a bit concerned that we get the correct address under weird configurations described in #17901 - maybe the protocol stack will set whatever address it has available?

File: src/common/address.c [hunk 23]
```diff
@@ -1404,8 +1407,16 @@ ifaddrs_to_smartlist(const struct ifaddrs *ifa, sa_family_t family)
       continue;
     if (family != AF_UNSPEC && i->ifa_addr->sa_family != family)
       continue;
+    if (loopback && (i->ifa_flags & IFF_LOOPBACK) == 0)
+      continue;
+    if (!loopback && (i->ifa_flags & IFF_LOOPBACK) == IFF_LOOPBACK)
+      continue;
```
[the remark this answers is not in the document]
Comment (rl1987) [reply src/common/address.c]: But that's what happens here?

File: src/common/address.c [hunk 24]
```diff
@@ -1569,18 +1591,33 @@ ifreq_to_smartlist(char *buf, size_t buflen)
...
+
+      if (valid_sa_family && conversion_success) {
+        if (tor_addr_is_null(&tmp))
+          continue;
+        if (tor_addr_is_multicast(&tmp))
+          continue;
+        if (loopback && !tor_addr_is_loopback(&tmp))
+          continue;
+        if (!loopback && tor_addr_is_loopback(&tmp))
+          continue;
```
[the remark this answers is not in the document]
Comment (rl1987) [reply src/common/address.c]: I looked into this. I think we're using ioctl() wrong. Will open a ticket. Meanwhile, this code will work as workaround.

File: src/common/address.c [hunk 25]
```diff
@@ -1648,21 +1687,23 @@ get_interface_addresses_ioctl(int severity, sa_family_t family)
  * (An empty smartlist indicates that we successfully learned that we have no
  * addresses.)  Log failure messages at <b>severity</b>. Only return the
  * interface addresses of requested <b>family</b> and ignore the addresses
- * of other address families. */
+ * of other address families. If <b>loopback</b> is true, retrieve
+ * only loopback interface addresses. */
```
[the remark this answers is not in the document]
Comment (rl1987) [reply src/common/address.c]: Fixed.
